# KHÁM PHÁ VÀ TIỀN XỬ LÝ DỮ LIỆU

---

## MỤC LỤC:
- [1. Các thư viện cần thiết](#phan-1)
- [2. Đọc dữ liệu lưu vào Dataframe](#phan-2)
- [3. Khám phá dữ liệu](#phan-3)
    - [1. Số dòng, số cột của tập dữ liệu](#phan-3-1)
    - [2. Ý nghĩa mỗi dòng/cột dữ liệu](#phan-3-2)
    - [3. Kiểm tra các dòng trùng lặp](#phan-3-3)
    - [4. Kiểu dữ liệu của các cột](#phan-3-4)
    - [5. Phân bố của các cột dữ liệu](#phan-3-5)
        - [5.1. Cột dữ liệu dạng số](#phan-3-5-1)
        - [5.2. Cột dữ liệu dạng phân loại](#phan-3-5-2)
            - [5.2.1. Cột có giá trị đơn](#phan-3-5-2-1)
            - [5.2.2. Cột có đa giá trị](#phan-3-5-2-2)
        - [5.3. Nhận xét chung](#phan-3-5-3)
- [4. Xuất file](#phan-4)

---

## 1. Các thư viện cần thiết: <a id="phan-1"></a>

In [1]:
import pandas as pd
import numpy as np

## 2. Đọc dữ liệu lưu vào Dataframe: <a id="phan-2"></a>

In [2]:
# Đọc tệp CSV mà không có header
netflix_df = pd.read_csv('../data/netflix_titles.csv', encoding='ISO-8859-1', skiprows= 0)
netflix_df.set_index('show_id', inplace=True) # Chọn cột show_id làm index
netflix_df.head()

type                  title         director  \
show_id                                                    
s1         Movie   Dick Johnson Is Dead  Kirsten Johnson   
s2       TV Show          Blood & Water              NaN   
s3       TV Show              Ganglands  Julien Leclercq   
s4       TV Show  Jailbirds New Orleans              NaN   
s5       TV Show           Kota Factory              NaN   

                                                      cast        country  \
show_id                                                                     
s1                                                     NaN  United States   
s2       Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
s3       Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
s4                                                     NaN            NaN   
s5       Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

                 date_added  release_year rating   duration  \
show_id                                                       
s1       September 25, 2021          2020  PG-13     90 min   
s2       September 24, 2021          2021  TV-MA  2 Seasons   
s3       September 24, 2021          2021  TV-MA   1 Season   
s4       September 24, 2021          2021  TV-MA   1 Season   
s5       September 24, 2021          2021  TV-MA  2 Seasons   

                                                 listed_in  ... Unnamed: 16  \
show_id                                                     ...               
s1                                           Documentaries  ...         NaN   
s2         International TV Shows, TV Dramas, TV Mysteries  ...         NaN   
s3       Crime TV Shows, International TV Shows, TV Act...  ...         NaN   
s4                                  Docuseries, Reality TV  ...         NaN   
s5       International TV Shows, Romantic TV Shows, TV ...  ...         NaN   

         Unnamed: 17  Unnamed: 18  Unnamed: 19  Unnamed: 20  Unnamed: 21  \
show_id                                                                    
s1               NaN          NaN          NaN          NaN          NaN   
s2               NaN          NaN          NaN          NaN          NaN   
s3               NaN          NaN          NaN          NaN          NaN   
s4               NaN          NaN          NaN          NaN          NaN   
s5               NaN          NaN          NaN          NaN          NaN   

         Unnamed: 22  Unnamed: 23  Unnamed: 24  Unnamed: 25  
show_id                                                      
s1               NaN          NaN          NaN          NaN  
s2               NaN          NaN          NaN          NaN  
s3               NaN          NaN          NaN          NaN  
s4               NaN          NaN          NaN          NaN  
s5               NaN          NaN          NaN          NaN  

[5 rows x 25 columns]

- Giải thích cho việc sử dụng cột `show_id` trong tập dữ liệu gốc làm index cho các dòng dữ liệu: Do các giá trị trong cột `show_id` là các giá trị độc lập, không trùng lắp nên sẽ thuận tiện cho việc truy xuất từng bộ phim/chương trình truyền hình thông qua ID của chúng.

## 3. Khám phá dữ liệu: <a id="phan-3"></a>

### 1. Số dòng, số cột của tập dữ liệu: <a id="phan-3-1"></a>

In [3]:
# Số dòng, số cột
row, column = netflix_df.shape
print(f'Số dòng của tập dữ liệu: {row}')
print(f'Số cột của tập dữ liệu: {column}')

Số dòng của tập dữ liệu: 8809
Số cột của tập dữ liệu: 25


- Sau khi quan sát, nhóm nhận thấy những cột cuối là những cột trống nên ở đây nhóm sẽ thực hiện xoá các cột trống.

In [4]:
# Xoá đi những cột có tất cả dòng có giá trị NaN
netflix_df = netflix_df.dropna(how='all', axis=1)
netflix_df.head()

type                  title         director  \
show_id                                                    
s1         Movie   Dick Johnson Is Dead  Kirsten Johnson   
s2       TV Show          Blood & Water              NaN   
s3       TV Show              Ganglands  Julien Leclercq   
s4       TV Show  Jailbirds New Orleans              NaN   
s5       TV Show           Kota Factory              NaN   

                                                      cast        country  \
show_id                                                                     
s1                                                     NaN  United States   
s2       Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
s3       Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
s4                                                     NaN            NaN   
s5       Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

                 date_added  release_year rating   duration  \
show_id                                                       
s1       September 25, 2021          2020  PG-13     90 min   
s2       September 24, 2021          2021  TV-MA  2 Seasons   
s3       September 24, 2021          2021  TV-MA   1 Season   
s4       September 24, 2021          2021  TV-MA   1 Season   
s5       September 24, 2021          2021  TV-MA  2 Seasons   

                                                 listed_in  \
show_id                                                      
s1                                           Documentaries   
s2         International TV Shows, TV Dramas, TV Mysteries   
s3       Crime TV Shows, International TV Shows, TV Act...   
s4                                  Docuseries, Reality TV   
s5       International TV Shows, Romantic TV Shows, TV ...   

                                               description  
show_id                                                     
s1       As her father nears the end of his life, filmm...  
s2       After crossing paths at a party, a Cape Town t...  
s3       To protect his family from a powerful drug lor...  
s4       Feuds, flirtations and toilet talk go down amo...  
s5       In a city of coaching centers known to train I...

In [5]:
# Số dòng, số cột sau khi loại bỏ các cột trống
row, column = netflix_df.shape
print(f'Số dòng của tập dữ liệu sau khi loại bỏ các cột trống: {row}')
print(f'Số cột của tập dữ liệu sau khi loại bỏ các cột trống: {column}')

Số dòng của tập dữ liệu sau khi loại bỏ các cột trống: 8809
Số cột của tập dữ liệu sau khi loại bỏ các cột trống: 11


### 2. Ý nghĩa mỗi dòng/cột dữ liệu: <a id="phan-3-2"></a>

#### Ý nghĩa mỗi dòng dữ liệu:

- Mỗi dòng dữ liệu cung cấp thông tin chi tiết về một bộ phim hoặc chương trình truyền hình, với các nội dung riêng biệt và rõ ràng.

#### Ý nghĩa mỗi cột dữ liệu:

- Mỗi cột dữ liệu cho biết những thông tin của một bộ phim hoặc chương trình truyền hình, bao gồm:
    - type: Danh mục có thể là bộ phim hoặc chương trình truyền hình.
    - title: Tựa đề của các bộ phim hoặc chương trình truyền hình.
    - director: Đạo diễn của các bộ phim/chương trình truyền hình. Một số mục sẽ có giá trị NULL, đặc biệt với các chương trình truyền hình.
    - cast: Danh sách các diễn viên chính.
    - country: Quốc gia là nơi sản xuất phim/chương trình truyền hình.
    - date_added: Ngày được đăng tải trên Netflix.
    - release_year: Năm sản xuất của bộ phim hoặc chương trình truyền hình.
    - rating: Quy định về độ tuổi được phép xem một nội dung xác định.
    - duration: Thời lượng của các bộ phim/chương trình truyền hình. Được đo bằng phút đối với các bộ phim và phần đối với các chương trình truyền hình.
    - listed_in: Thể loại mà tiêu đề thuộc về.
    - description: Mô tả ngắn về bộ phim/chương trình truyền hình.

### 3. Kiểm tra các dòng trùng lặp: <a id="phan-3-3"></a>

In [6]:
# Kiểm tra có tồn tại các dòng trùng lặp
row_duplicates = netflix_df[netflix_df.duplicated()]
print(f'Số dòng bị trùng lặp: {len(row_duplicates)}')

Số dòng bị trùng lặp: 3


- Như vậy trong tập dữ liệu có chứa các dòng trùng lặp.
- Tiến hành xoá các dòng trùng lặp.

In [7]:
# Thực hiện xoá các dòng trùng lặp
netflix_df = netflix_df.drop_duplicates()

In [8]:
# Kiểm tra còn tồn tại các dòng trùng lặp hay không
row_duplicates = netflix_df[netflix_df.duplicated()]
print(f'Số dòng bị trùng lặp sau khi thực hiện xoá: {len(row_duplicates)}')

Số dòng bị trùng lặp sau khi thực hiện xoá: 0


- Không còn tồn tại dòng nào bị trùng dữ liệu.

### 4. Kiểu dữ liệu của các cột: <a id="phan-3-4"></a>

In [9]:
# Kiểm tra kiểu dữ liệu của từng cột
netflix_df.dtypes

type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object

- Có thể thấy giá trị của các cột gần như đã đúng hết. Nhưng để tiện sử dụng cho sau này thì nhóm sẽ thực hiện chuyển kiểu dữ liệu của cột dữ liệu `date_added` thành `datetime64` và cột `duration` vừa là đo bằng phút đối với các bộ phim vừa là phần đối với các chương trình truyền hình nên cần xử lý tách thành 2 cột riêng biệt để đảm bảo từng cột có một đơn vị đo khác nhau.

In [10]:
# Chuyển cột date_added thành kiểu datetime 64
netflix_df['date_added'] = netflix_df['date_added'].str.strip() # Chuyển thành chuỗi để loại bỏ khoảng trắng ở đầu và cuối của chuỗi thời gian
netflix_df['date_added'] = pd.to_datetime(netflix_df['date_added'])

In [11]:
# Tách cột duration thành 2 cột riêng biệt
movie_df = netflix_df[netflix_df['type'] == 'Movie']
movie_df = movie_df['duration'].str.extract(r'(\d+)\s*(\w+)')[0].rename('duration1')
tv_df = netflix_df[netflix_df['type'] == 'TV Show']
tv_df = tv_df['duration'].str.extract(r'(\d+)\s*(\w+)')[0].rename('season')
netflix_df = netflix_df.join([tv_df,movie_df])
netflix_df

type                  title         director  \
show_id                                                    
s1         Movie   Dick Johnson Is Dead  Kirsten Johnson   
s2       TV Show          Blood & Water              NaN   
s3       TV Show              Ganglands  Julien Leclercq   
s4       TV Show  Jailbirds New Orleans              NaN   
s5       TV Show           Kota Factory              NaN   
...          ...                    ...              ...   
s8805      Movie             Zombieland  Ruben Fleischer   
s8806      Movie                   Zoom     Peter Hewitt   
s8807      Movie                 Zubaan      Mozez Singh   
s8808    TV Show     Parasyte: The Grey     Yeon Sang-ho   
s8809      Movie                 Serena     Susanne Bier   

                                                      cast        country  \
show_id                                                                     
s1                                                     NaN  United States   
s2       Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
s3       Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
s4                                                     NaN            NaN   
s5       Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
...                                                    ...            ...   
s8805    Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
s8806    Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
s8807    Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   
s8808           Shin Hyun-been, Jeon Yeo-bin, Goo Kyo-hwan    South Korea   
s8809    Bradley Cooper, Jennifer Lawrence, Toby Jones,...  United States   

        date_added  release_year rating   duration  \
show_id                                              
s1      2021-09-25          2020  PG-13     90 min   
s2      2021-09-24          2021  TV-MA  2 Seasons   
s3      2021-09-24          2021  TV-MA   1 Season   
s4      2021-09-24          2021  TV-MA   1 Season   
s5      2021-09-24          2021  TV-MA  2 Seasons   
...            ...           ...    ...        ...   
s8805   2019-11-01          2009      R     88 min   
s8806   2020-01-11          2006     PG     88 min   
s8807   2019-03-02          2015  TV-14    111 min   
s8808   2024-04-05          2024  TV-MA   1 Season   
s8809   2024-04-05          2014      A    110 min   

                                                 listed_in  \
show_id                                                      
s1                                           Documentaries   
s2         International TV Shows, TV Dramas, TV Mysteries   
s3       Crime TV Shows, International TV Shows, TV Act...   
s4                                  Docuseries, Reality TV   
s5       International TV Shows, Romantic TV Shows, TV ...   
...                                                    ...   
s8805                              Comedies, Horror Movies   
s8806                   Children & Family Movies, Comedies   
s8807       Dramas, International Movies, Music & Musicals   
s8808                               Sci-fi, Horror, Action   
s8809                             Drama, Romance, Thriller   

                                               description season duration1  
show_id                                                                      
s1       As her father nears the end of his life, filmm...    NaN        90  
s2       After crossing paths at a party, a Cape Town t...      2       NaN  
s3       To protect his family from a powerful drug lor...      1       NaN  
s4       Feuds, flirtations and toilet talk go down amo...      1       NaN  
s5       In a city of coaching centers known to train I...      2       NaN  
...                                                    ...    ...       ...  
s8805    Looking to survive in a world taken over by zo...    NaN        88  
s8806    Dragged 

In [12]:
# Xoá cột duration ban đầu
netflix_df = netflix_df.drop('duration', axis=1)
# Đổi tên cột duration1 tức là cột duration mới
netflix_df = netflix_df.rename(columns={'duration1': 'duration'})

In [13]:
# Kiểm tra lại kiểu dữ liệu của từng cột
netflix_df.dtypes

type                    object
title                   object
director                object
cast                    object
country                 object
date_added      datetime64[ns]
release_year             int64
rating                  object
listed_in               object
description             object
season                  object
duration                object
dtype: object

- Sau khi tách cột duration ban đầu thành 2 cột: 1 cột season dành cho số phần của các chương trình truyền hình và 1 cột duration mới dành cho thời lượng của các bộ phim (đơn vị: Phút)
- Nhưng kiểu dữ liệu ở 2 cột vẫn chưa phù hợp với ngữ nghĩa, cần chuyển thành kiểu `float64` do có tồn tại giá trị `NaN`.
    - Giải thích cho giá trị `NaN` có tồn tại trong 2 cột dữ liệu: Do đối với các bộ phim thì cột season sẽ có giá trị `NaN` vì season chỉ dành cho các bộ phim truyền hình, và ngược lại.

In [14]:
# Chuyển cột 'duration' và 'season' thành kiểu float
netflix_df['duration'] = netflix_df['duration'].astype('float')
netflix_df['season'] = netflix_df['season'].astype('float')

In [15]:
# Kiểm tra lại kiểu dữ liệu của từng cột
netflix_df.dtypes

type                    object
title                   object
director                object
cast                    object
country                 object
date_added      datetime64[ns]
release_year             int64
rating                  object
listed_in               object
description             object
season                 float64
duration               float64
dtype: object

- Như vậy, các cột dữ liệu đã có kiểu dữ liệu phù hợp với ngữ nghĩa của chúng.

### 5. Phân bố của các cột dữ liệu: <a id="phan-3-5"></a>

#### 5.1. Cột dữ liệu dạng số: <a id="phan-3-5-1"></a>

In [16]:
# Lấy ra các cột numerical
numeric_columns = netflix_df.select_dtypes(include='number').columns[:]
numeric_columns

Index(['release_year', 'season', 'duration'], dtype='object')

Ta xem xét tỉ lệ giá trị bị thiếu trong các cột:

In [17]:
percentage_missing_dic = {}
for col in numeric_columns:
    percentage_missing = netflix_df[col].isna().mean() * 100
    percentage_missing_dic[col] = percentage_missing
    print(f'Phần trăm giá trị còn thiếu của cột \'{col}\' là {percentage_missing:.2f}%')

Phần trăm giá trị còn thiếu của cột 'release_year' là 0.00%
Phần trăm giá trị còn thiếu của cột 'season' là 69.61%
Phần trăm giá trị còn thiếu của cột 'duration' là 30.42%


Vì trong dữ liệu, phim điện ảnh chỉ có thời lượng (số phút), phim truyền hình chỉ có số mùa, nên khi tách ra 2 cột sẽ tạo giá trị `Null`. Với cột season, giá trị null đến từ phim điện ảnh, nó không có số mùa. Với cột duration, giá trị null đến từ phim truyền hình, nó không có thời lượng cụ thể. Để đảm bảo không đưa ra các giả định không cần thiết khi số mùa không đóng vai trò quan trọng với phim điện ảnh và thời lượng không đóng vai trò quan trọng đối với chương trình truyền hình nên những giá trị null được giữ lại.

**Tiếp theo ta xem qua mô tả của các cột có dữ liệu dạng số xem có gì bất thường không (tối thiểu, tối đa)**

In [18]:
netflix_df.describe()

date_added  release_year       season     duration
count                           8796   8806.000000  2676.000000  6127.000000
mean   2019-05-17 16:36:59.918144512   2014.179877     1.764948    99.567651
min              2008-01-01 00:00:00   1925.000000     1.000000     3.000000
25%              2018-04-06 00:00:00   2013.000000     1.000000    87.000000
50%              2019-07-02 12:00:00   2017.000000     1.000000    98.000000
75%              2020-08-20 00:00:00   2019.000000     2.000000   114.000000
max              2024-04-05 00:00:00   2024.000000    17.000000   312.000000
std                              NaN      8.820097     1.582752    28.285800

Ta xem xét từng cột và nhận xét:
- release_year có giá trị từ 1925 đến 2024, là thời gian năm phù hợp.
- season có giá trị từ 1 đến 17, là số mùa phù hợp cho các phim truyền hình. Số mùa trung bình là 1 cho thấy rằng max khá xa so với giá trị trung bình.
- duration có giá trị từ 3 đến 312 phút, có sự chênh lệch khá lớn.
- Các cột đều có khoảng giá trị phù hợp với ý nghĩa từng cột.

#### 5.2. Cột dữ liệu dạng phân loại: <a id="phan-3-5-2"></a>

Ở phần này sẽ có 2 dạng:
- Cột có giá trị đơn: type, title, director, date_added, rating, description.
- Cột có nhiều giá trị: cast, country, listed_in.

##### 5.2.1. Cột có giá trị đơn: <a id="phan-3-5-2-1"></a>

**Xử lý cột giá trị đơn**

In [19]:
# Lấy các cột đơn giá trí
single_columns = ['type', 'title', 'director', 'date_added', 'rating', 'description']

In [20]:
# Tính giá trị thiếu cho các cột đơn giá trị
for col in single_columns:
    percentage_missing = netflix_df[col].isna().mean() * 100
    percentage_missing_dic[col] = percentage_missing
    print(f'Phần trăm giá trị còn thiếu của cột \'{col}\' là {percentage_missing:.2f}%')

Phần trăm giá trị còn thiếu của cột 'type' là 0.00%
Phần trăm giá trị còn thiếu của cột 'title' là 0.00%
Phần trăm giá trị còn thiếu của cột 'director' là 29.90%
Phần trăm giá trị còn thiếu của cột 'date_added' là 0.11%
Phần trăm giá trị còn thiếu của cột 'rating' là 0.05%
Phần trăm giá trị còn thiếu của cột 'description' là 0.00%


Ta in các dòng bị thiếu dữ liệu của các biến để xem xét

- `director`:

In [21]:
# Kiểm tra director
netflix_df[netflix_df['director'].isna()]

type                                title director  \
show_id                                                          
s2       TV Show                        Blood & Water      NaN   
s4       TV Show                Jailbirds New Orleans      NaN   
s5       TV Show                         Kota Factory      NaN   
s11      TV Show  Vendetta: Truth, Lies and The Mafia      NaN   
s15      TV Show      Crime Stories: India Detectives      NaN   
...          ...                                  ...      ...   
s8796    TV Show                      Yu-Gi-Oh! Arc-V      NaN   
s8797    TV Show                           Yunus Emre      NaN   
s8798    TV Show                            Zak Storm      NaN   
s8801    TV Show                   Zindagi Gulzar Hai      NaN   
s8804    TV Show                          Zombie Dumb      NaN   

                                                      cast  \
show_id                                                      
s2       Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   
s4                                                     NaN   
s5       Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...   
s11                                                    NaN   
s15                                                    NaN   
...                                                    ...   
s8796    Mike Liscio, Emily Bauer, Billy Bob Thompson, ...   
s8797    Gökhan Atalay, Payidar Tüfekçioglu, Baran A...   
s8798    Michael Johnston, Jessica Gee-George, Christin...   
s8801    Sanam Saeed, Fawad Khan, Ayesha Omer, Mehreen ...   
s8804                                                  NaN   

                                               country date_added  \
show_id                                                             
s2                                        South Africa 2021-09-24   
s4                                                 NaN 2021-09-24   
s5                                               India 2021-09-24   
s11                                                NaN 2021-09-24   
s15                                                NaN 2021-09-22   
...                                                ...        ...   
s8796                                    Japan, Canada 2018-05-01   
s8797                                           Turkey 2017-01-17   
s8798    United States, France, South Korea, Indonesia 2018-09-13   
s8801                                         Pakistan 2016-12-15   
s8804                                              NaN 2019-07-01   

         release_year rating  \
show_id                        
s2               2021  TV-MA   
s4               2021  TV-MA   
s5               2021  TV-MA   
s11              2021  TV-MA   
s15              2021  TV-MA   
...               ...    ...   
s8796            2015  TV-Y7   
s8797            2016  TV-PG   
s8798            2016  TV-Y7   
s8801            2012  TV-PG   
s8804            2018  TV-Y7   

                                                 listed_in  \
show_id                                                      
s2         International TV Shows, TV Dramas, TV Mysteries   
s4                                  Docuseries, Reality TV   
s5       International TV Shows, Romantic TV Shows, TV ...   
s11      Crime TV Shows, Docuseries, International TV S...   
s15           British TV Shows, Crime TV Shows, Docuseries   
...                                                    ...   
s8796                               Anime Series, Kids' TV   
s8797                    International TV Shows, TV Dramas   
s8798                                             Kids' TV   
s8801    International TV Shows, Romantic TV Shows, TV ...   
s8804               Kids' TV, Korean TV Shows, TV Comedies   

                                               description  season  duration  
show_id                                                                       
s2       After crossing paths at a party, a Cape Town t...     2.0

Vì có thể có 1 số nội dung không cần đạo diễn (ví dụ: tài liệu ngắn hoặc show truyền hình thực tế) và dữ liệu thiếu cũng không nhỏ (29.9%) nên để có thể giữ lại được dữ liệu, nhóm sẽ điền giá trị UNKNOWN

In [22]:
netflix_df['director'] = netflix_df['director'].fillna('UNKNOWN')

- `date-added`:

In [23]:
# Kiểm tra date_added
netflix_df[netflix_df['date_added'].isna()]

type                                        title director  \
show_id                                                                  
s6067    TV Show  A Young Doctor's Notebook and Other Stories  UNKNOWN   
s6175    TV Show              Anthony Bourdain: Parts Unknown  UNKNOWN   
s6796    TV Show                                      Frasier  UNKNOWN   
s6807    TV Show                                      Friends  UNKNOWN   
s6902    TV Show                              Gunslinger Girl  UNKNOWN   
s7197    TV Show                                     Kikoriki  UNKNOWN   
s7255    TV Show                          La Familia P. Luche  UNKNOWN   
s7407    TV Show                                        Maron  UNKNOWN   
s7848    TV Show                                 Red vs. Blue  UNKNOWN   
s8183    TV Show                 The Adventures of Figaro Pho  UNKNOWN   

                                                      cast         country  \
show_id                                                                      
s6067    Daniel Radcliffe, Jon Hamm, Adam Godley, Chris...  United Kingdom   
s6175                                     Anthony Bourdain   United States   
s6796    Kelsey Grammer, Jane Leeves, David Hyde Pierce...   United States   
s6807    Jennifer Aniston, Courteney Cox, Lisa Kudrow, ...   United States   
s6902    Yuuka Nanri, Kanako Mitsuhashi, Eri Sendai, Am...           Japan   
s7197                                        Igor Dmitriev             NaN   
s7255    Eugenio Derbez, Consuelo Duval, Luis Manuel Á...   United States   
s7407    Marc Maron, Judd Hirsch, Josh Brener, Nora Zeh...   United States   
s7848    Burnie Burns, Jason Saldaña, Gustavo Sorola, ...   United States   
s8183    Luke Jurevicius, Craig Behenna, Charlotte Haml...       Australia   

        date_added  release_year rating  \
show_id                                   
s6067          NaT          2013  TV-MA   
s6175          NaT          2018  TV-PG   
s6796          NaT          2003  TV-PG   
s6807          NaT          2003  TV-14   
s6902          NaT          2008  TV-14   
s7197          NaT          2010   TV-Y   
s7255          NaT          2012  TV-14   
s7407          NaT          2016  TV-MA   
s7848          NaT          2015     NR   
s8183          NaT          2015  TV-Y7   

                                                 listed_in  \
show_id                                                      
s6067             British TV Shows, TV Comedies, TV Dramas   
s6175                                           Docuseries   
s6796                       Classic & Cult TV, TV Comedies   
s6807                       Classic & Cult TV, TV Comedies   
s6902                         Anime Series, Crime TV Shows   
s7197                                             Kids' TV   
s7255    International TV Shows, Spanish-Language TV Sh...   
s7407                                          TV Comedies   
s7848    TV Action & Adventure, TV Comedies, TV Sci-Fi ...   
s8183                                Kids' TV, TV Comedies   

                                               description  season  duration  
show_id                                                                       
s6067    Set during the Russian Revolution, this comic ...     2.0       NaN  
s6175    This CNN original series has chef Anthony Bour...     5.0       NaN  
s6796    Frasier Crane is a snooty but lovable Seattle ...    11.0       NaN  
s6807    This hit sitcom follows the merry misadventure...    10.0       NaN  
s6902    On the surface, the Social Welfare Agency appe...     2.0       NaN  
s7197    A wacky rabbit and his gang of animal pals hav...     2.0       NaN  
s7255    This irreverent sitcom featues Ludovico, Feder...     3.0       NaN  
s7407    Marc Maron stars as Marc Maron, who interviews...     4.0       NaN  
s7848    This parody of first-person shooter games, mil...    13.0       NaN  
s8183    Imagine your worst fears, then multiply them: 

Vì tỷ lệ giá trị thiếu của `date_added` khá thấp và thời gian không dễ suy luận hoặc dự đoán chính xác, nếu điền sẽ dễ làm sai lệch phân bố thời gian, nên nhóm sẽ loại bỏ những hàng có giá trị thời gian thiếu.

In [24]:
netflix_df = netflix_df.dropna(subset=['date_added'])

- `rating`:

In [25]:
# Kiểm tra rating
netflix_df[netflix_df['rating'].isna()]

type                                              title  \
show_id                                                               
s5990      Movie  13TH: A Conversation with Oprah Winfrey & Ava ...   
s6828    TV Show                  Gargantia on the Verdurous Planet   
s7313    TV Show                                       Little Lunch   
s7538      Movie                               My Honor Was Loyalty   

                director                                               cast  \
show_id                                                                       
s5990            UNKNOWN                        Oprah Winfrey, Ava DuVernay   
s6828            UNKNOWN  Kaito Ishikawa, Hisako Kanemoto, Ai Kayano, Ka...   
s7313            UNKNOWN  Flynn Curry, Olivia Deeble, Madison Lu, Oisín...   
s7538    Alessandro Pepe  Leone Frisa, Paolo Vaccarino, Francesco Miglio...   

           country date_added  release_year rating  \
show_id                                              
s5990          NaN 2017-01-26          2017    NaN   
s6828        Japan 2016-12-01          2013    NaN   
s7313    Australia 2018-02-01          2015    NaN   
s7538        Italy 2017-03-01          2015    NaN   

                                    listed_in  \
show_id                                         
s5990                                  Movies   
s6828    Anime Series, International TV Shows   
s7313                   Kids' TV, TV Comedies   
s7538                                  Dramas   

                                               description  season  duration  
show_id                                                                       
s5990    Oprah Winfrey sits down with director Ava DuVe...     NaN      37.0  
s6828    After falling through a wormhole, a space-dwel...     1.0       NaN  
s7313    Adopting a child's perspective, this show take...     1.0       NaN  
s7538    Amid the chaos and horror of World War II, a c...     NaN     115.0

Vì số lượng ít nên có thể tìm nhãn phim để gán vào.

In [26]:
netflix_df.loc[netflix_df.index=='s5990','rating'] = 'T13'
netflix_df.loc[netflix_df.index=='s6828','rating'] = 'TV-14'
netflix_df.loc[netflix_df.index=='s7313','rating'] = 'TV-G'
netflix_df.loc[netflix_df.index=='s7538','rating'] = 'PG-13'


**Kiểm tra các loại trong giá trị đơn**

In [27]:
# Lưu số lượng các giá trị khác nhau và đếm số lượng từng loại single_cate_profile
single_cate_profile = pd.DataFrame(index=["num_diff_vals", "distribution"])
num_diff_vals = netflix_df[single_columns].nunique()
distribution = netflix_df[single_columns].agg(lambda column: pd.Series(column).value_counts().to_dict())
single_cate_profile[single_columns] = [num_diff_vals, distribution]

single_cate_profile

type  \
num_diff_vals                                 2   
distribution   {'Movie': 6130, 'TV Show': 2666}   

                                                           title  \
num_diff_vals                                               8796   
distribution   {'Dick Johnson Is Dead': 1, 'Care Bears & Cous...   

                                                        director  \
num_diff_vals                                               4530   
distribution   {'UNKNOWN': 2623, 'Rajiv Chilaka': 19, 'Raúl ...   

                                                      date_added  \
num_diff_vals                                               1715   
distribution   {2020-01-01 00:00:00: 110, 2019-11-01 00:00:00...   

                                                          rating  \
num_diff_vals                                                 19   
distribution   {'TV-MA': 3206, 'TV-14': 2156, 'TV-PG': 861, '...   

                                                     description  
num_diff_vals                                               8767  
distribution   {'Paranormal activity at a lush, abandoned pro...

In [28]:
single_cate_profile.loc['distribution','rating']

{'TV-MA': 3206,
 'TV-14': 2156,
 'TV-PG': 861,
 'R': 798,
 'PG-13': 491,
 'TV-Y7': 333,
 'TV-Y': 306,
 'PG': 287,
 'TV-G': 221,
 'NR': 79,
 'G': 41,
 'TV-Y7-FV': 6,
 'NC-17': 3,
 'UR': 3,
 '74 min': 1,
 '84 min': 1,
 '66 min': 1,
 'T13': 1,
 'A': 1}

Vì có 1 số giá trị rating không phù hợp như: 66 min, 84 min, 74 min, và chúng không nhiều nên nhóm sẽ tìm và thay thế chúng cho phù hợp.

In [29]:
netflix_df[netflix_df['rating'].isin(['66 min', '84 min', '74 min'])]

type                                 title    director        cast  \
show_id                                                                        
s5542    Movie                       Louis C.K. 2017  Louis C.K.  Louis C.K.   
s5795    Movie                 Louis C.K.: Hilarious  Louis C.K.  Louis C.K.   
s5814    Movie  Louis C.K.: Live at the Comedy Store  Louis C.K.  Louis C.K.   

               country date_added  release_year  rating listed_in  \
show_id                                                             
s5542    United States 2017-04-04          2017  74 min    Movies   
s5795    United States 2016-09-16          2010  84 min    Movies   
s5814    United States 2016-08-15          2015  66 min    Movies   

                                               description  season  duration  
show_id                                                                       
s5542    Louis C.K. muses on religion, eternal love, gi...     NaN       NaN  
s5795    Emmy-winning comedy writer Louis C.K. brings h...     NaN       NaN  
s5814    The comic puts his trademark hilarious/thought...     NaN       NaN

In [30]:
netflix_df.loc[netflix_df.index=='s5542', 'rating'] = 'TV-MA'
netflix_df.loc[netflix_df.index=='s5795', 'rating'] = 'Not Rated'
netflix_df.loc[netflix_df.index=='s5814', 'rating'] = 'Not Rated'

In [31]:
single_cate_profile

type  \
num_diff_vals                                 2   
distribution   {'Movie': 6130, 'TV Show': 2666}   

                                                           title  \
num_diff_vals                                               8796   
distribution   {'Dick Johnson Is Dead': 1, 'Care Bears & Cous...   

                                                        director  \
num_diff_vals                                               4530   
distribution   {'UNKNOWN': 2623, 'Rajiv Chilaka': 19, 'Raúl ...   

                                                      date_added  \
num_diff_vals                                               1715   
distribution   {2020-01-01 00:00:00: 110, 2019-11-01 00:00:00...   

                                                          rating  \
num_diff_vals                                                 19   
distribution   {'TV-MA': 3206, 'TV-14': 2156, 'TV-PG': 861, '...   

                                                     description  
num_diff_vals                                               8767  
distribution   {'Paranormal activity at a lush, abandoned pro...

- Sau khi chỉnh sửa lại thì các cột đều có giá trị không có gì bất thường.
- Có tổng cộng 2 loại hình, 8796 tiêu đề, 4529 đạo diễn, 1715 ngày thêm, 19 rating, 8767 mô tả trừ những giá trị UNKNOWN.

##### 5.2.2. Cột có đa giá trị: <a id="phan-3-5-2-2"></a>

**Xử lý cột đa giá trị**

In [32]:
# Lấy các cột đa giá trị
multi_columns = ['cast', 'country', 'listed_in']

In [33]:
# Tính giá trị thiếu của các cột đa giá trị
for col in multi_columns:
    percentage_missing = netflix_df[col].isna().mean() * 100
    percentage_missing_dic[col] = percentage_missing
    print(f'Phần trăm giá trị còn thiếu của cột \'{col}\' là {percentage_missing:.2f}%')

Phần trăm giá trị còn thiếu của cột 'cast' là 9.38%
Phần trăm giá trị còn thiếu của cột 'country' là 9.42%
Phần trăm giá trị còn thiếu của cột 'listed_in' là 0.00%


Ta thấy các cột này có giá trị thiếu không ít. Có thể có những nội dung không có diễn viên hay quốc gia cụ thể, để có thể giữ được dữ liệu, ta điền vào dữ liệu thiếu giá trị 'UNKNOWN'

In [34]:
netflix_df['cast'] = netflix_df['cast'].fillna('UNKNOWN')
netflix_df['country'] = netflix_df['country'].fillna('UNKNOWN')

In [35]:
# Lưu số lượng giá trị khác nhau và đếm số lượng từng loại và lưu vào mảng

# Tạo dataframe kết quả
def split_and_melt(column):
    df_split = column.str.split(', ', expand=True).melt().drop('variable', axis=1)
    return df_split['value']
long_df = netflix_df[multi_columns].agg(split_and_melt)
distribution = long_df.agg(lambda column: pd.Series(column).value_counts().to_dict())
num_diff_vals = long_df.nunique()

# Lưu và đếm
multiple_cate_profile = pd.DataFrame(index=["num_diff_vals", "distribution"])
multiple_cate_profile[multi_columns] = [num_diff_vals, distribution]

multiple_cate_profile

cast  \
num_diff_vals                                              36431   
distribution   {'UNKNOWN': 825, 'Anupam Kher': 43, 'Shah Rukh...   

                                                         country  \
num_diff_vals                                                128   
distribution   {'United States': 3683, 'India': 1045, 'UNKNOW...   

                                                       listed_in  
num_diff_vals                                                 48  
distribution   {'International Movies': 2752, 'Dramas': 2425,...

Ta có thể trích xuất cột thể loại để xem thử

In [36]:
multiple_cate_profile.loc['distribution','listed_in']

{'International Movies': 2752,
 'Dramas': 2425,
 'Comedies': 1673,
 'International TV Shows': 1349,
 'Documentaries': 869,
 'Action & Adventure': 859,
 'TV Dramas': 761,
 'Independent Movies': 755,
 'Children & Family Movies': 641,
 'Romantic Movies': 616,
 'Thrillers': 576,
 'TV Comedies': 574,
 'Crime TV Shows': 469,
 "Kids' TV": 449,
 'Docuseries': 394,
 'Music & Musicals': 375,
 'Romantic TV Shows': 370,
 'Horror Movies': 357,
 'Stand-Up Comedy': 343,
 'Reality TV': 255,
 'British TV Shows': 252,
 'Sci-Fi & Fantasy': 243,
 'Sports Movies': 219,
 'Anime Series': 175,
 'Spanish-Language TV Shows': 173,
 'TV Action & Adventure': 167,
 'Korean TV Shows': 151,
 'Classic Movies': 116,
 'LGBTQ Movies': 102,
 'TV Mysteries': 98,
 'Science & Nature TV': 92,
 'TV Sci-Fi & Fantasy': 83,
 'TV Horror': 75,
 'Cult Movies': 71,
 'Anime Features': 71,
 'Teen TV Shows': 69,
 'Faith & Spirituality': 65,
 'Movies': 57,
 'TV Thrillers': 57,
 'Stand-Up Comedy & Talk Shows': 56,
 'Classic & Cult TV': 26

Dữ liệu không có giá trị bất thường.

Có tổng cộng 36430 diễn viên, 127 quốc gia, và 48 thể loại trừ những giá trị UNKNOWN.

#### 5.3. Nhận xét chung: <a id="phan-3-5-3"></a>

- Các giá trị của các cột và các hàng sau khi tiền xử lý không có gì bất thường, các con số đều phù hợp với ý nghĩa của từng cột dữ liệu.
- Dữ liệu thiếu ở cột duration và season là do việc phân loại, những hàng có loại không quan trọng thuộc tính này thì nó sẽ để null.
- Đa số các cột có dữ liệu thiếu ít.
- Các cột cũng có các khoảng giá trị phù hợp tên cột.

## 4. Xuất file: <a id="phan-4"></a>

In [37]:
netflix_df.to_csv('netflix_titles_cleaned.csv', index=False)